src/mscorlib/shared/System/StringComparer.cs

```diff
@@ -244,18 +283,72 @@ public override int GetHashCode(string obj)
                 throw new ArgumentNullException(nameof(obj));
 #endif
             }
+            Contract.EndContractBlock();
+
+            if (_ignoreCase)
+            {
+                return TextInfo.GetHashCodeOrdinalIgnoreCase(obj);
+            }
+
             return obj.GetHashCode();
         }
 
-        // Equals/GetHashCode methods for the comparer itself. 
-        public override bool Equals(object obj) => obj is OrdinalComparer;
-        public override int GetHashCode() => nameof(OrdinalComparer).GetHashCode();
+        // Equals method for the comparer itself. 
+        public override bool Equals(object obj)
+        {
+            OrdinalComparer comparer = obj as OrdinalComparer;
+            if (comparer == null)
+            {
+                return false;
+            }
+            return (this._ignoreCase == comparer._ignoreCase);
+        }
+
+        public override int GetHashCode()
+        {
+            int hashCode = nameof(OrdinalComparer).GetHashCode();
+            return _ignoreCase ? (~hashCode) : hashCode;
+        }
     }
 
     [Serializable]
-    [System.Runtime.CompilerServices.TypeForwardedFrom("mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089")]
-    internal sealed class OrdinalIgnoreCaseComparer : StringComparer
+    internal sealed class OrdinalCaseSensitiveComparer : OrdinalComparer, ISerializable
+    {
+        public OrdinalCaseSensitiveComparer() : base(false)
+        {
+        }
+
+        public override int Compare(string x, string y) => string.CompareOrdinal(x, y);
+
+        public override bool Equals(string x, string y) => string.Equals(x, y);
+
+        public override int GetHashCode(string obj)
+        {
+            if (obj == null)
+            {
+#if CORECLR
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.obj);
+#else
+                throw new ArgumentNullException(nameof(obj));
+#endif
```

Comment:
> Do we ever compile this code without CORECLR set

Yes, we do - in CoreRT. Regular exception throwing works better in ProjectN (because of the toolchain recognizes and does special optimizations for) than the ThrowHelper hack (that the toolchain does not recognize). It makes folks to sometime put the ThrowHelper under ifdefs when moving code into the shared corelib partition. We are not consistent about it though. I think that the ifdef can be removed in this case without any harm.

Related:
https://github.com/dotnet/roslyn/issues/15929
https://github.com/dotnet/corert/issues/1832